### DISCUSSION TITLE
CI: Travis, Cirrus or GitHub Actions?

### DISCUSSION THREAD

**Discussion by AlexanderLanin** [ccache/ccache#611] [Development]
### Summary ###
Besides Travis CI there are many other CI systems. Obviously a change should not happen for fun, so it must have important advantages. On first glance Cirrus CI qualifies as a significant improvement. However before investing too much time I'm interested in hearing feedback.

### Details ###
All:
* Free for open source. Exact limits / runtimes to be investigated.
* supports Linux and MacOS

Pro Travis CI:
* Works fine
* Effort for switch (setup, new syntax, etc)

Pro Cirrus CI:
* Somewhat similar syntax with yml file
* Support Linux, macOS, **Windows** and **FreeBSD**, it seems they fully support **docker** too.
* Supports arbitrary **caching** (e.g. docker files, ccache cache)

Pro GitHub Actions:
* Somewhat similar syntax with yml file
* Support Linux, macOS and **Windows**, it seems they fully support **docker** too.
* Probably the **best possible GitHub integration**
* It seems they have a rather high limit on free parallel jobs of 20 (travis runs 3 dual core jobs in parallel?!)

**Comment by jrosdahl** [comment]
How does CI using GitHub Actions compare?

**Comment by AlexanderLanin** [comment]
Fair question. If we look into one we should look into more then one.

There is no "best CI". I don't know whether they are just not comparable or the "best CI" changes every few months. But you'll find as many opinions as websites you open.

I've added GitHub Actions to the text on top. Based on that I guess GitHub actions in the best. While it would be really nice to run FreeBSD, the GitHub integration of GitHub actions is a huge advantage for simplifying contributions to ccache

Looking at this trend travis has difficult times ahead. Not even mentioning cirrus.
https://trends.google.de/trends/explore?date=today%205-y&q=cirrus%20ci,travis%20ci,github%20actions

**Comment by cristianadam** [comment]
I've blogged about using GitHub Actions with C++, CMake and ccache at: https://cristianadam.eu/20200113/speeding-up-c-plus-plus-github-actions-using-ccache/

I had a GitHub Actions yml file for ccache at my fork https://github.com/cristianadam/ccache/commit/dbac3d7d2f7b4452dad9f67ac9a53ef7be62c11f

It doesn't run the bash tests. But since GitHub Actions provide bash on all platforms, it could be used to run the tests.

Qt6 QtBase has a bash setup https://github.com/qt/qtbase/blob/dev/.github/workflows/ninja-build.yml

**Comment by AlexanderLanin** [comment]
Thanks, I've taken a slightly different approach by first replicating exactly what is available in travis ci today just to verify it's the same kind of complexity: https://github.com/ccache/ccache/pull/621/files

I'm not sure about maturity of GitHub Actions, but having artifacts is a great advantage (not only compiled binaries, but also testdirs from operating systems you don't have yourself!). It's 2-3x faster than Travis (10 ->4 minutes).

**Comment by jrosdahl** [comment]
> Thanks, I've taken a slightly different approach by first replicating exactly what is available in travis ci today

Excellent.

> I'm not sure about maturity of GitHub Actions

Me neither, but I'm getting very tired of Travis-CI since it stops working all the time.

**Comment by jrosdahl** [comment]
See #621.
